add SNS connector with publish sink #204

fg-devs opened the pull request:
- Includes a new connector called sns
- Includes a sink to publish messages to an AWS SNS topic

johanandren commented:
REFS #204

johanandren commented on sns/src/main/scala/akka/stream/alpakka/sns/SnsPublishSinkStage.scala:
```diff
@@ -0,0 +1,77 @@
+
+import scala.concurrent.{Future, Promise}
+
+final class SnsPublishSinkStage(topicArn: String, snsClient: AmazonSNSAsync)
```
Is the `AmazonSNSAsync` thread safe? A stage can be materialized any number of times and all of them will share the same `snsClient`.  If it cannot be shared a factory lambda is better (`() => AmazonSNSAsync`), that way every materialization can have its own client (which it will be responsible to manage/close etc in sync with the stage lifecycle).

Another option might be to take a `AmazonSNSAsyncClientBuilder` and `.build()` that each materialization (if the builder is immutable that is).

johanandren replied:
An additional thought: It could be better to make this a flow, emitting the values downstream as they have been ack:ed from AWS, the factory for the simple use case as sink could then be achieved through composition (`SNSPublishFlow.to(Sink.ignore)`).

fg-devs replied:
Yes the AmazonSNSAsync is ThreadSafe regarding to the AWS [docs.](http://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/services/sns/AmazonSNSAsyncClient.html) That's why I think it is fine to share a single AmazonSNSAsync among multiple materialization stages.

Also regarding your last comment about not making the AmazonSNSAsync implicit, I try to be consistent with other connectors, like the SQS one. There we also pass the AmazonSQSClient implicitly. I think it will be easier for people who use the connectors if the initialization of the GraphStages work the same way.

Good point to change the Sink to a Flow. I thought about keeping the Sink and provide an additional Flow GraphStage which will return the messageId of the published message, so clients can do further processing with the returned messageId.

johanandren replied:
Ah, Hadn't noticed that how many connectors that do that. Thats a valid argument to keep it implicit.

johanandren replied:
Thanks for verifying that the client is threadsafe.

johanandren replied:
Maybe even just let it emit PublishResult downstream and then both String => String, and String => MessageId can be achieved through combinations with other stages.

fg-devs replied:
Updated the SnsPublishFlow to emit a PublishResult.

johanandren replied:
This means we could dlete the sink stage and express it by using the flow in the factory method instead:
```
val flow: Flow[String, PublishResult, NotUsed] 
val sink: Sink[String, Future[Done]] = flow.toMat(Sink.ignore)(Keep.right)
```

johanandren commented on sns/src/main/scala/akka/stream/alpakka/sns/SnsPublishSinkStage.scala:
```diff
@@ -0,0 +1,77 @@
+      }
+
+      def handleFailure(ex: Throwable): Unit = {
+        log.error(ex, "AmazonSNSAsync failure: {}", ex.getMessage)
```
No need to log the failure since we fail the stage and the materialized value with it. You could wrap it with a en exception to decorate with extra information from the stage though.

fg-devs replied:
Alright I will remove it then.

johanandren commented on sns/src/test/scala/akka/stream/alpakka/sns/SnsPublishSinkSpec.scala:
```diff
@@ -0,0 +1,125 @@
+              publishRequest,
+              new PublishResult().withMessageId("message-id")
+            )
+          new CompletableFuture()
```
This future is never completed though, shouldn't that cause the `matVal` to never complete, is there a bug around waiting for ack on publishing?

fg-devs replied:
I will further investigate it.

johanandren commented on sns/src/main/scala/akka/stream/alpakka/sns/scaladsl/SnsPublishSink.scala:
```diff
@@ -0,0 +1,18 @@
+  /**
+   * Scala API: creates a [[SnsPublishSinkStage]] for a SNS topic using an [[AmazonSNSAsync]]
+   */
+  def apply(topicArn: String)(implicit snsClient: AmazonSNSAsync): Sink[String, Future[Done]] =
```
I don't think the client should be implicit, especially considering the note about thread safety.

fg-devs commented on build.sbt:
```diff
@@ -77,6 +77,14 @@ lazy val simpleCodecs = project
   )
 
+lazy val sns = project
+  .in(file("sns"))
```
Alright, I will try it out.

fg-devs commented on docs/src/main/paradox/sns.md:
```diff
@@ -0,0 +1,77 @@
+# AWS SNS Connector
+
+The AWS SNS connector provides an Akka Stream sink for AWS SNS.
```
Good point, I will add some more information.

fg-devs commented:
@johanandren I updated the PR regarding your comments. Furthemrore, I also added a SnsPublishFlow what I mentioned in one of my replies of your comments.

johanandren commented on sns/src/main/scala/akka/stream/alpakka/sns/SnsPublishFlowStage.scala:
```diff
@@ -0,0 +1,59 @@
+        val request = new PublishRequest().withTopicArn(topicArn).withMessage(grab(in))
+
+        snsClient.publishAsync(request,
+          new AsyncHandler[PublishRequest, PublishResult] {
```
Not necessarily anything to act on but, as this gets the values passed in on success/failure, it could be a single instance of `AsyncHandler` that you reuse rather than a new instance for every message.

johanandren commented on sns/src/main/scala/akka/stream/alpakka/sns/SnsPublishFlowStage.scala:
```diff
@@ -0,0 +1,59 @@
+      }
+
+      override def onPull(): Unit = {
+        if (isClosed(in)) completeStage()
```
There could be an ongoing publish that did not yet complete here, even if in was closed, you need to keep track of that and emit that last publish result when it comes, and then complete the stage.

johanandren commented on sns/src/main/scala/akka/stream/alpakka/sns/javadsl/SnsPublishSink.scala:
```diff
@@ -0,0 +1,26 @@
+
+import scala.compat.java8.FutureConverters._
+
+object SnsPublishSink {
```
Maybe this should then instead be `SnsPublisher` and have both `.sink` and `.flow` in the same place?

fg-devs commented:
@johanandren Updated regarding your comments. Removed the SnsPublishSink and provide a SnsPublisher with .flow and .sink factory methods. Also implemented a way to keep track of in flight messages for proper stage closing.

patriknw commented on sns/src/main/scala/akka/stream/alpakka/sns/SnsPublishFlowStage.scala:
```diff
@@ -0,0 +1,62 @@
+        log.debug("Published SNS message: {}", result.getMessageId)
+        inFlight -= 1
+        if (isAvailable(out)) {
+          if (!hasBeenPulled(in)) tryPull(in)
```
why `pull` here? shouldn't that only be done from `onPull`?

fg-devs replied:
Wouldn't this limit the Flow to process elements only one by one? Otherwise we would always have to wait for the output port to pull?

Thats why I also made inFlight as an Int and not a Boolean.

patriknw commented on sns/src/main/scala/akka/stream/alpakka/sns/SnsPublishFlowStage.scala:
```diff
@@ -0,0 +1,62 @@
+      private def handleSuccess(result: PublishResult): Unit = {
+        log.debug("Published SNS message: {}", result.getMessageId)
+        inFlight -= 1
+        if (isAvailable(out)) {
```
when will it not be `available`? should this be `!isClosed` 
we are just dropping the message if it's not `available` so I want to be sure that we don't do that accidentally

patriknw commented on sns/src/main/scala/akka/stream/alpakka/sns/SnsPublishFlowStage.scala:
```diff
@@ -0,0 +1,62 @@
+      }
+
+      override def onPush(): Unit = {
+        inFlight += 1
```
can `inFlight` be > 1? perhaps use `boolean` if it can only be 0 or 1

patriknw commented on sns/src/main/scala/akka/stream/alpakka/sns/SnsPublishFlowStage.scala:
```diff
@@ -0,0 +1,62 @@
+
+      override def onPull(): Unit = {
+        if (isClosed(in) && inFlight == 0) completeStage()
+        if (!hasBeenPulled(in)) tryPull(in)
```
`else if`

patriknw commented on sns/src/main/scala/akka/stream/alpakka/sns/SnsPublishFlowStage.scala:
```diff
@@ -0,0 +1,62 @@
+import com.amazonaws.services.sns.AmazonSNSAsync
+import com.amazonaws.services.sns.model.{PublishRequest, PublishResult}
+
+final class SnsPublishFlowStage(topicArn: String, snsClient: AmazonSNSAsync)
```
Internal API?
`private[akka]`

patriknw commented on sns/src/main/scala/akka/stream/alpakka/sns/javadsl/SnsPublisher.scala:
```diff
@@ -0,0 +1,28 @@
+object SnsPublisher {
+
+  /**
+   * Java API: creates a [[Flow]] from a [[SnsPublishFlowStage]] for a SNS topic using an [[AmazonSNSAsync]]
```
users should not care about the implementation `SnsPublishFlowStage`

ktoso commented:
Hi there @fg-devs, are you going to have some time to address the small comments made?
I'd hope so, would be awesome to include this in the next release :)

fg-devs commented:
@ktoso Hey, yes I will do the changes today.

fg-devs commented:
@ktoso @patriknw Update the PR regarding your comments. The build for Scala 2.12 failed (there is already an open [ticket)](https://github.com/akka/alpakka/issues/235)

johanandren commented on sns/src/main/scala/akka/stream/alpakka/sns/SnsPublishFlowStage.scala:
```diff
@@ -0,0 +1,59 @@
+import com.amazonaws.services.sns.AmazonSNSAsync
+import com.amazonaws.services.sns.model.{PublishRequest, PublishResult}
+
+private[akka] final class SnsPublishFlowStage(topicArn: String, snsClient: AmazonSNSAsync)
```
Does the client have a lifecycle, is is thread safe, should this be a client factory instead (`() => AmazonSNSAsync`) and have its lifecycle managed by the stage (`client.close() when the stage stops)?

johanandren replied:
NVM, I already asked that previously but github hid all comments and I had forgot.
